File: AI-Reply/Shared/Account/APIClient.swift
import Foundation

/// Every way a backend call can fail, as a value the UI can localize.
///
/// Сервер қатесі — тұрақты код, аударма қосымшада.
///
/// The server sends a stable `code`; the English `message` beside it is for a
/// developer reading a log, never for a user. Mapping happens once, here, so no
/// screen has to know what an HTTP status means.
enum APIError: Error, Equatable, Sendable {
    case offline
    case timedOut
    case cancelled
    /// The session is gone: the user has to sign in again.
    case unauthorized
    case accountDisabled
    case invalidOTP
    case otpExpired
    case rateLimited(retryAfter: Int?)
    /// Daily quota is spent. `resetsAt` is ISO-8601 from the server.
    case dailyLimitReached(limit: Int, usedToday: Int, resetsAt: String?)
    case subscriptionExpired
    case paymentRequired
    case providerUnavailable
    case providerTimeout
    case emptyResponse
    case invalidRequest
    case notFound
    case conflict
    case server
    /// The client could not make sense of the response at all.
    case malformedResponse

    /// Whether retrying the same request could plausibly succeed.
    var isTransient: Bool {
        switch self {
        case .offline, .timedOut, .providerUnavailable, .providerTimeout, .server: return true
        default: return false
        }
    }
}

/// Thin HTTP client for the AI Reply backend.
///
/// No third-party dependency, no retry loops hidden inside, no shared mutable
/// state: one method, one request, one decoded value.
struct APIClient: Sendable {

    private let baseURL: URL
    private let session: URLSession

    init(baseURL: URL, session: URLSession? = nil) {
        self.baseURL = baseURL
        self.session = session ?? ReplyNetworking.makeSession(timeout: AIConfiguration.requestTimeout)
    }

    /// Decoded response plus the bearer token that was used, so a caller that
    /// refreshed mid-flight does not have to ask again.
    struct Empty: Decodable, Sendable {}

    // MARK: Requests

    func get<Response: Decodable>(_ path: String, token: String? = nil) async throws -> Response {
        try await send(path: path, method: "GET", body: Optional<Empty>.none, token: token)
    }

    @discardableResult
    func post<Body: Encodable, Response: Decodable>(
        _ path: String, body: Body, token: String? = nil
    ) async throws -> Response {
        try await send(path: path, method: "POST", body: body, token: token)
    }

    @discardableResult
    func patch<Body: Encodable, Response: Decodable>(
        _ path: String, body: Body, token: String? = nil
    ) async throws -> Response {
        try await send(path: path, method: "PATCH", body: body, token: token)
    }

    // MARK: Transport

    private func send<Body: Encodable, Response: Decodable>(
        path: String, method: String, body: Body?, token: String?
    ) async throws -> Response {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONEncoder().encode(body)
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw Self.mapTransportError(error)
        }

        guard let http = response as? HTTPURLResponse else { throw APIError.malformedResponse }

        guard (200..<300).contains(http.statusCode) else {
            throw Self.mapServerError(status: http.statusCode, data: data, headers: http)
        }

        if Response.self == Empty.self { return Empty() as! Response }

        do {
            return try JSONDecoder().decode(Response.self, from: data)
        } catch {
            throw APIError.malformedResponse
        }
    }

    // MARK: Error mapping

    /// The error envelope every endpoint uses.
    private struct ErrorEnvelope: Decodable {
        struct Payload: Decodable {
            let code: String
            let message: String?
            let details: Details?
        }
        struct Details: Decodable {
            let dailyLimit: Int?
            let usedToday: Int?
            let resetsAt: String?
            let retryAfterSeconds: Int?

            enum CodingKeys: String, CodingKey {
                case dailyLimit = "daily_limit"
                case usedToday = "used_today"
                case resetsAt = "resets_at"
                case retryAfterSeconds = "retry_after_seconds"
            }
        }
        let error: Payload
    }

    static func mapServerError(status: Int, data: Data, headers: HTTPURLResponse) -> APIError {
        let envelope = try? JSONDecoder().decode(ErrorEnvelope.self, from: data)
        let details = envelope?.error.details
        let retryAfter = details?.retryAfterSeconds
            ?? Int(headers.value(forHTTPHeaderField: "Retry-After") ?? "")

        switch envelope?.error.code {
        case "UNAUTHORIZED", "TOKEN_EXPIRED":   return .unauthorized
        case "ACCOUNT_DISABLED":                return .accountDisabled
        case "INVALID_OTP":                     return .invalidOTP
        case "OTP_EXPIRED":                     return .otpExpired
        case "RATE_LIMITED":                    return .rateLimited(retryAfter: retryAfter)
        case "DAILY_LIMIT_REACHED", "MONTHLY_LIMIT_REACHED":
            return .dailyLimitReached(limit: details?.dailyLimit ?? 0,
                                      usedToday: details?.usedToday ?? 0,
                                      resetsAt: details?.resetsAt)
        case "SUBSCRIPTION_EXPIRED":            return .subscriptionExpired
        case "PAYMENT_REQUIRED":                return .paymentRequired
        case "AI_PROVIDER_UNAVAILABLE":         return .providerUnavailable
        case "AI_TIMEOUT":                      return .providerTimeout
        case "AI_EMPTY_RESPONSE":               return .emptyResponse
        case "INVALID_REQUEST":                 return .invalidRequest
        case "NOT_FOUND":                       return .notFound
        case "CONFLICT":                        return .conflict
        default: break
        }

        // No envelope (a proxy error page, say): fall back to the status.
        switch status {
        case 401, 403: return .unauthorized
        case 404:      return .notFound
        case 408, 504: return .providerTimeout
        case 409:      return .conflict
        case 429:      return .rateLimited(retryAfter: retryAfter)
        case 400...499: return .invalidRequest
        default:       return .server
        }
    }

    static func mapTransportError(_ error: Error) -> APIError {
        if error is CancellationError { return .cancelled }
        let nsError = error as NSError
        guard nsError.domain == NSURLErrorDomain else { return .server }
        switch nsError.code {
        case NSURLErrorCancelled: return .cancelled
        case NSURLErrorTimedOut:  return .timedOut
        case NSURLErrorNotConnectedToInternet,
             NSURLErrorNetworkConnectionLost,
             NSURLErrorCannotFindHost,
             NSURLErrorCannotConnectToHost,
             NSURLErrorDataNotAllowed,
             NSURLErrorInternationalRoamingOff:
            return .offline
        default: return .server
        }
    }
}

File: AI-Reply-Android/app/src/main/java/kz/yerek/aireply/data/account/ApiClient.kt
package kz.yerek.aireply.data.account

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.IOException
import java.net.ConnectException
import java.net.HttpURLConnection
import java.net.NoRouteToHostException
import java.net.SocketTimeoutException
import java.net.URL
import java.net.UnknownHostException
import javax.net.ssl.SSLException
import kotlin.coroutines.coroutineContext

/**
 * Every way a backend call can fail, as a value the UI can localize.
 *
 * Сервер қатесі — тұрақты код, аудармасы қосымшада.
 *
 * The server sends a stable `code`; the English `message` next to it is for a
 * developer reading a log, never for a user. Mapping happens once, here, so no
 * screen has to know what an HTTP status means.
 */
sealed interface ApiError {
    data object Offline : ApiError
    data object TimedOut : ApiError
    data object Cancelled : ApiError

    /** The session is gone: the user has to sign in again. */
    data object Unauthorized : ApiError
    data object AccountDisabled : ApiError
    data object InvalidOtp : ApiError
    data object OtpExpired : ApiError
    data class RateLimited(val retryAfterSeconds: Int?) : ApiError

    /** Daily quota is spent. [resetsAt] is ISO-8601 from the server. */
    data class DailyLimitReached(
        val limit: Int,
        val usedToday: Int,
        val resetsAt: String?
    ) : ApiError

    data object SubscriptionExpired : ApiError
    data object PaymentRequired : ApiError
    data object ProviderUnavailable : ApiError
    data object ProviderTimeout : ApiError
    data object EmptyResponse : ApiError
    data object InvalidRequest : ApiError
    data object NotFound : ApiError
    data object Conflict : ApiError
    data object Server : ApiError

    /** The client could not make sense of the response at all. */
    data object MalformedResponse : ApiError
}

/** Thrown across suspend boundaries; the payload is what the UI actually reads. */
class ApiException(val error: ApiError) : Exception(error::class.simpleName)

fun ApiError.raise(): Nothing = throw ApiException(this)

/**
 * Thin HTTP client for the AI Reply backend.
 *
 * WHY HttpURLConnection, AGAIN. Same reason as [kz.yerek.aireply.ai.ReplyNetworking]:
 * this package is reachable from the input method, and a networking library
 * there is thousands of classes loaded while the user stares at a blank key
 * strip. Four verbs, one JSON codec, no interceptors.
 *
 * PATCH is deliberately absent: HttpURLConnection refuses it outright, so the
 * profile update goes through the server's POST alias instead of a reflection
 * hack on a platform class.
 */
class ApiClient(
    private val baseUrl: String,
    private val timeoutMs: Int = DEFAULT_TIMEOUT_MS
) {

    val json: Json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        explicitNulls = false
    }

    /**
     * Performs one request and returns the body, or throws [ApiException].
     *
     * The response body of a FAILED request is read for the error envelope only
     * and never logged: it can quote the request back, and a request can carry
     * a private message.
     */
    suspend fun request(
        method: String,
        path: String,
        body: String? = null,
        token: String? = null
    ): String = withContext(Dispatchers.IO) {
        val url = baseUrl.trimEnd('/') + "/" + path.trimStart('/')
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = timeoutMs
            readTimeout = timeoutMs
            useCaches = false
            setRequestProperty("Accept", "application/json")
            if (body != null) setRequestProperty("Content-Type", "application/json; charset=utf-8")
            if (token != null) setRequestProperty("Authorization", "Bearer $token")
            doOutput = body != null
        }

        // Cancellation is real, not cooperative-only: the socket read is
        // unblocked as soon as the calling coroutine goes away.
        val disconnectOnCancel = coroutineContext[Job]?.invokeOnCompletion { cause ->
            if (cause != null) runCatching { connection.disconnect() }
        }

        try {
            if (body != null) {
                connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            }

            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val text = stream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }.orEmpty()

            if (status !in 200..299) {
                throw ApiException(
                    mapServerError(status, text, connection.getHeaderField("Retry-After"))
                )
            }
            text
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (api: ApiException) {
            throw api
        } catch (throwable: Throwable) {
            throw ApiException(mapTransportError(throwable))
        } finally {
            disconnectOnCancel?.dispose()
            runCatching { connection.disconnect() }
        }
    }

    companion object {
        const val DEFAULT_TIMEOUT_MS = 25_000

        private val envelopeJson = Json { ignoreUnknownKeys = true }

        /** Stable server codes first; the HTTP status only as a fallback. */
        fun mapServerError(status: Int, body: String, retryAfterHeader: String? = null): ApiError {
            val payload = runCatching {
                envelopeJson.decodeFromString<ErrorEnvelopeDto>(body).error
            }.getOrNull()
            val details = payload?.details
            val retryAfter = details?.retryAfterSeconds ?: retryAfterHeader?.toIntOrNull()

            when (payload?.code) {
                "UNAUTHORIZED", "TOKEN_EXPIRED" -> return ApiError.Unauthorized
                "ACCOUNT_DISABLED" -> return ApiError.AccountDisabled
                "INVALID_OTP" -> return ApiError.InvalidOtp
                "OTP_EXPIRED" -> return ApiError.OtpExpired
                "RATE_LIMITED" -> return ApiError.RateLimited(retryAfter)
                "DAILY_LIMIT_REACHED", "MONTHLY_LIMIT_REACHED" -> return ApiError.DailyLimitReached(
                    limit = details?.dailyLimit ?: 0,
                    usedToday = details?.usedToday ?: 0,
                    resetsAt = details?.resetsAt
                )
                "SUBSCRIPTION_EXPIRED" -> return ApiError.SubscriptionExpired
                "PAYMENT_REQUIRED" -> return ApiError.PaymentRequired
                "AI_PROVIDER_UNAVAILABLE" -> return ApiError.ProviderUnavailable
                "AI_TIMEOUT" -> return ApiError.ProviderTimeout
                "AI_EMPTY_RESPONSE" -> return ApiError.EmptyResponse
                "INVALID_REQUEST" -> return ApiError.InvalidRequest
                "NOT_FOUND" -> return ApiError.NotFound
                "CONFLICT" -> return ApiError.Conflict
            }

            // No envelope: a proxy error page, or the old /v1 shape.
            return when (status) {
                401, 403 -> ApiError.Unauthorized
                404 -> ApiError.NotFound
                408, 504 -> ApiError.ProviderTimeout
                409 -> ApiError.Conflict
                413 -> ApiError.InvalidRequest
                429 -> ApiError.RateLimited(retryAfter)
                in 400..499 -> ApiError.InvalidRequest
                else -> ApiError.Server
            }
        }

        fun mapTransportError(throwable: Throwable): ApiError = when (throwable) {
            is CancellationException -> ApiError.Cancelled
            is SocketTimeoutException -> ApiError.TimedOut
            is UnknownHostException, is ConnectException, is NoRouteToHostException -> ApiError.Offline
            is SSLException -> ApiError.Server
            is IOException -> ApiError.Server
            else -> ApiError.Server
        }
    }
}
